tests/integration: Enable "images serve: check zeroconf" test in CI plus minor improvements

lucas-p-bernardes opened the pull request:
Pipeline: https://gitlab.int.toradex.com/lucas.bernardes/torizoncore-builder/-/pipelines/118698

lucas-p-bernardes committed:
gitignore: Add workdir directory, to avoid tracking temp test files

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
gitlab-ci: Remove ./setup.sh and export TCBCMD of .test-base job script

This was already being done by run_all.sh step, that overwrite this two
steps defined in job itself, so they are useless.
Otherwise, since test-torizoncore-common doesn't use run_all.sh, this is
still needed, so they are defined only for common now.

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
gitlab-ci: Remove creation of workdir/reports dir in test-base job

This is already created in run.sh, so creating this in job itself is
useless.

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
functions.sh: Print torizoncore-builder-bg alias definition if test fail

This is is helpfull to debug tests wich use this function.

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
setup.sh: Remove unnecessary ; in tcb_clean_storage_volume

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes committed:
tests/integration: Enable "images serve: check zeroconf" test in CI

To run the avahi-browse-domains command, it was necessary to use an Alpine
container, which configures DBUS, starts the avahi-daemon in the background,
and finally runs avahi-browse-domains.

Additionally, when running the tests with dind configured to use a TCP port (as
is the case with this repo), Avahi was unable to discover services on the
network with --network=host. On the other hand, when dind is mounted using
docker.sock, Avahi was able to find network services, indicating that there is
a conflict  between dind running with the TCP port and Avahi. To solve this, a
Docker network tcb_network was created, where the command torizoncore-builder
images serve is first run in background, and then avahi-browse-domains is
executed, making it possible to discover services on the network in this
Docker network.

Related-to: TCB-503

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes commented:
Hello @rborn-tx @jsrc27, can we have this merged?

rborn-tx commented on tests/integration/functions.sh:
```diff
@@ -42,10 +42,20 @@ torizoncore-builder-bg() {
+    unset TCBCMD_NETWORK
+
+    # Print alias definition if test fail
+    echo "torizoncore-builder-bg alias was defined as: $CMD $@"
```
This `torizoncore-builder-bg` is a generic function that can be used in many tests and this new stretch of code looks too specific IMO. A better design would be to make it even more generic and keep the specific parts somewhere else.

Here's a way to do that:

```
torizoncore-builder-bg() {
    local tcbcmd
    if [ -z "${TCBCMD_BG_OVERRIDE}" ]; then
	tcbcmd="${TCBCMD}"
    else
	tcbcmd="${TCBCMD_BG_OVERRIDE}"
    fi
    # And replace TCBCMD -> tcbcmd in the rest of the function
    ...
}
```

The other function that starts TCB with a specific network would then do something like this:

```
torizoncore-builder-bg-tcb_network() {
    local tcbcmd="${TCBCMD/--network=host/--network=tcb_network}"
    TCBCMD_BG_OVERRIDE="${tcbcmd}" torizoncore-builder-bg "$@"
}
export -f torizoncore-builder-bg-tcb_network
```

Notice we don't need any global variable in this version (the value of `${TCBCMD_BG_OVERRIDE}` is passed only to `torizoncore-builder-bg`) which is another improvement.

---

As a second point, notice that `torizoncore-builder-bg-tcb_network` is coupled with `tcb_create_docker_network` which is in another module (`setup.sh`) and the connection between the two is not visible. If someone decides to change the name of the network, things would simply stop working and one would need to figure out why. It would be better to keep the name of the network in some variable exported by `setup.sh` say `TCB_BG_ALT_NETWORK` and use that variable in both functions.

In this case, I'd rename `torizoncore-builder-bg-tcb_network` to say `torizoncore-builder-bg-alt-network` to indicate that this new function will start TCB on the background on an "alternative" network.

lucas-p-bernardes replied:
Done!

rborn-tx commented on tests/integration/testcases/images-serve.bats:
```diff
@@ -24,12 +24,17 @@ teardown() {
-    run avahi-browse-domains -a -t
+    torizoncore-builder-bg-tcb_network images serve $IMAGE_DIR
+
+    run docker run -i --rm --privileged --network=tcb_network alpine:latest sh -c "
```
Here again we'd use `${TCB_BG_ALT_NETWORK}`.

lucas-p-bernardes replied:
Done!

rborn-tx reviewed (changes requested):
Sorry for the delay - I overlooked this PR. I have some suggestions for better design/code reuse.

lucas-p-bernardes commented:
Hello @rborn-tx, I fixed your points, and rebase the branch. Please take a look if everything was fixed correctly and merge if it's OK.
This is the pipeline with this last changes: https://gitlab.int.toradex.com/lucas.bernardes/torizoncore-builder/-/jobs/937051

lucas-p-bernardes commented:
@rborn-tx The tests start failing in Github (and only in Github) when I defined the name of the docker network using the variable `TCB_BG_ALT_NETWORK`, defined in the `tcb_create_docker_network()` script in `setup.sh`, because the variable  is not persisting across steps, since each `run:` command in GitHub Actions is executed in a new shell, meaning that exported variables do not persist between steps.
Do you have a suggestion for fix this?

rborn-tx commented:
> @rborn-tx The tests start failing in Github (and only in Github) when I defined the name of the docker network using the variable `TCB_BG_ALT_NETWORK`, defined in the `tcb_create_docker_network()` script in `setup.sh`, because the variable is not persisting across steps, since each `run:` command in GitHub Actions is executed in a new shell, meaning that exported variables do not persist between steps. Do you have a suggestion for fix this?

Looking at the other steps I see the environment variables are passed between steps by appending to a file referenced by variable `GITHUB_ENV`. Then the step that does the setup would become:

```
      - name: Run setup script
        run: |
          TCB_SKIP_PULL=1 ./setup.sh
          echo "TCB_BG_ALT_NETWORK=${TCB_BG_ALT_NETWORK}" >> ${GITHUB_ENV}
```

In the future we could think of something more elaborate such as asking the setup script itself to dump the exported variables.

lucas-p-bernardes committed:
Github Actions: Persist TCB_BG_ALT_NETWORK across workflow steps

Previously, TCB_BG_ALT_NETWORK was only set within the setup script's
execution, making it unavailable in subsequent workflow steps. This change
appends the variable to ${GITHUB_ENV}, ensuring it persists across steps in
the GitHub Actions workflow.

Signed-off-by: Lucas Bernardes <lucas.bernardes@toradex.com>

lucas-p-bernardes commented:
@rborn-tx The bug of Github Actions failing was solved!

rborn-tx reviewed (approved):
LGTM
